content/the-basics-of-go.md

```diff
@@ -0,0 +1,864 @@
+then make some function from that.
+
+```go
+type Person struct {
+  Name string
+}
+
+func (p Person) SayHello() {
+  fmt.Println("Hello, my name is", p.Name)
+}
+```
+
+## Interface
+
+Interfaces is basically a set of method signatures. You can define an explicit interface of a type, or you can
+define an interface with a set of methods. Let's make an interface called `Speaker`.
+
+```go
+type Speaker interface {
+  SayHello()
+  SayGoodbye()
+}
+```
+
+You can make a struct that implements the interface. So, this is kind of what you'd expect when you're asking
+"oh no, I love OOP, why don't Go has OOP?". Listen here you little brat, OOP doesn't depend on whether the language
+have a "class" keyword or not. OOP is just a way of thinking about things. Not merely a specific way of doing things.
+
+Okay, let's end that rant. Let's make a struct called `Person` that implements the `Speaker` interface.
+
+```go
+type Person struct {
+  Name string
+}
+
+func (p Person) SayHello() {
+  fmt.Println("Hello, my name is", p.Name)
+}
+
+func (p Person) SayGoodbye() {
+  fmt.Println("Goodbye, my name is", p.Name)
+}
+
+func main() {
+  var speaker = Speaker
+  speaker = Person{Name: "John Doe"}
+  speaker.SayHello()
+  speaker.SayGoodbye()
+}
+```
+
+Interface could be empty. So, you can make an interface that has no methods. This is the equivalent of `any` (that
+means literally a type of anything), and might be useful when you're dealing with things that you don't know.
+
+But is it couraged to do so? No. You should be careful with interfaces. Go is not a dynamically typed language.
+So you have to validate the given type of the interface.
+
+```go
+func main() {
+	var i interface{}
+
+	i = 42
+	describe(i)
+
+	i = "hello"
+	describe(i)
+}
+
+func describe(i interface{}) {
+	fmt.Printf("(%v, %T)\n", i, i)
+  // Output:
+  // (42, int)
+  // (hello, string)
+}
+```
+
+To validate a given type, there are two ways. One, via the type assertion:
+
+```go
+var i interface{} = "hello"
+
+s := i.(string)
+fmt.Println(s)
+
+s, ok := i.(string)
+fmt.Println(s, ok)
+
+f, ok := i.(float64)
+fmt.Println(f, ok)
+```
+
+Or two, via the type switch:
+
+```go
+func do(i interface{}) {
+	switch v := i.(type) {
+	case int:
+		fmt.Printf("Twice %v is %v\n", v, v*2)
+	case string:
+		fmt.Printf("%q is %v bytes long\n", v, len(v))
+	default:
+		fmt.Printf("I don't know about type %T!\n", v)
+	}
+}
+
+func main() {
+	do(21)
+	do("hello")
+	do(true)
+}
+```
+
+## Error
+
+I want to tell a little story before we go forward. Go's error handling is a bit unique. It's not like C/C++,
+where you have to define a custom error type. But it's not like Java, where you have to define a custom exception.
+There are no try-catches. And there are no exceptions. But Go has an error type that must be defined explicitly,
+which is called `error`, and means your code will be so verbose. I once had an internal talk on one of the
```

Comment:
Go treat error as a value, m8 sound good.

Reply:
I talk about it on the paragraph below that

```
because in Go, error are just values. You can decorate the error,
and find out what was going on and where the error is without trying to read the stack trace.
```

Reply:
okeh, nice resolved